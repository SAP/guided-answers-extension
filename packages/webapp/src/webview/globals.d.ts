interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
