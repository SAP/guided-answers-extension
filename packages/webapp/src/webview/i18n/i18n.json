{
    "GUIDED_ANSWER_HEADER": "Guide",
    "SEARCH_GUIDED_ANSWERS": "Search Guided Answers",
    "ALL_ANSWERS": "Home",
    "STEP_BACK": "Step back",
    "RESTART": "Restart",
    "GUIDED_ANSWERS": "Guided Answers",
    "NO_ANSWERS_FOUND": "No answers found",
    "PLEASE_MODIFY_SEARCH": "Please modfiy search",
    "THIS_SOLVED_MY_ISSUE": "This solved my issue",
    "THIS_DID_NOT_SOLVED_MY_ISSUE": "This did not solve my issue",
    "THANK_YOU_TEXT": "We are glad to hear that your issue has been resolved and we hope you enjoy using the Guided Answers application.",
    "THANKS": "Thanks!",
    "HOME": "Home",
    "SEND": "Send",
    "CLOSE": "Close",
    "ISSUE_IS_NOT_RESOLVED": "Issue is not resolved",
    "WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED": "We are sorry to hear that your issue is not yet resolved.",
    "PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL": "Please tell us if this answer was helpful",
    "THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE": "There are several options for getting further assistance:",
    "SEARCH_FOR_ANOTHER_GUIDED_ANSWER": "Search for another Guided Answer",
    "START_AN_EXPERT_CHAT": "Start an Expert Chat",
    "SCHUEDULE_AN_EXPERT": "Schedule an Expert",
    "OPEN_AN_INCIDENT": "Open an Incident",
    "ASK_THE_SAP_COMMUNITY": "Ask the SAP Community",
    "FEEDBACK": "Feedback",
    "FEEDBACK_DIALOG_TITLE": "Is this content helpful?",
    "FEEDBACK_DIALOG_SUBTEXT": "If you have suggestions on how to improve this content we would love to hear them!",
    "FEEDBACK_DIALOG_DISCLAIMER": "Your feedback is anonymous, we do not collect any personal data.",
    "FEEDBACK_DIALOG_SUGGESTION": "Your suggestion",
    "FEEDBACK_SENT_DIALOG_TITLE": "Message sent",
    "FEEDBACK_SENT_DIALOG_SUBTEXT": "Thanks for the feedback!"
}
