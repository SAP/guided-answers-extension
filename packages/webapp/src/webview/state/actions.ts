export {
    updateGuidedAnswerTrees,
    updateActiveNode,
    selectNode,
    goToPreviousPage,
    goToAllAnswers,
    restartAnswer,
    executeCommand,
    setActiveTree,
    searchTree,
    setQueryValue,
    webviewReady,
    getBetaFeatures,
    sendFeedbackOutcome,
    sendFeedbackComment,
    guideFeedback,
    setProductFilters,
    setComponentFilters,
    resetFilters,
    setPageSize,
    feedbackStatus,
    feedbackResponse
} from '@sap/guided-answers-extension-types';
