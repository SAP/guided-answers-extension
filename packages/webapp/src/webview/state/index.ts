export { actions, store } from './store';
export { getInitialState } from './reducers';
