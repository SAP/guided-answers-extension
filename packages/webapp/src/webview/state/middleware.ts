import i18next from 'i18next';
import type { Middleware, MiddlewareAPI, Dispatch, Action } from 'redux';
import type { GuidedAnswerActions } from '@sap/guided-answers-extension-types';
import type { AppState } from '../types';

declare let window: Window;
declare let acquireVsCodeApi: () => typeof window['vscode'];

/**
 * Communication between IDE extension and web view is realized through the communication middleware
 * window event listener for incoming messages in to the webview
 * post message for outgoing messages to the host.
 *
 * @param store - redux store
 * @returns - redux middleware
 */
export const communicationMiddleware: Middleware<
    Dispatch<GuidedAnswerActions>,
    AppState,
    Dispatch<GuidedAnswerActions>
> = (store: MiddlewareAPI) => {
    // Add event handler, this will dispatch incomming state updates
    window.addEventListener('message', (event: MessageEvent) => {
        if (event.origin === window.origin) {
            console.log(i18next.t('MESSAGE_RECEIVED'), event);
            if (event.data && typeof event.data.type === 'string') {
                store.dispatch(event.data);
            }
        } else {
            console.error(i18next.t('UNKNOWN_ORIGIN') + event.origin);
        }
    });
    try {
        window.vscode = acquireVsCodeApi();
    } catch (e) {
        console.error(i18next.t('ERROR_NOT_RUNNING_IN_WEB_VIEW'));
    }
    return (next: Dispatch<GuidedAnswerActions>) =>
        (action): Action => {
            action = next(action);
            if (action && typeof action.type === 'string' && !action.type.startsWith('[view]')) {
                window.vscode.postMessage(action);
                console.log(i18next.t('REACT_ACTION_POSTED'), action);
            }
            return action;
        };
};
