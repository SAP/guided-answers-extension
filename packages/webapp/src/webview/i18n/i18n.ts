import { use } from 'i18next';
import { initReactI18next } from 'react-i18next';
import * as i18nEn from './i18n.json';

export { i18n } from 'i18next';

export function initI18n(): void {
    use(initReactI18next)
        .init({
            resources: {
                en: {
                    translation: i18nEn
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            interpolation: {
                escapeValue: false
            }
        })
        .catch((error) => console.error(error));
}
