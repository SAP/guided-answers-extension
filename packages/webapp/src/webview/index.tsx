import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
