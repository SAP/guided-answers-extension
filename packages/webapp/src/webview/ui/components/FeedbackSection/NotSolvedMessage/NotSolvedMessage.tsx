import React, { ReactElement } from 'react';
import i18next from 'i18next';
import '../../GuidedAnswerNode/GuidedAnswerNode.scss';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';
import { actions } from '../../../../state';

const options = [
    { link: 'https://launchpad.support.sap.com/#/expertchat/create', text: 'START_AN_EXPERT_CHAT' },
    { link: 'https://launchpad.support.sap.com/#/sae', text: 'SCHUEDULE_AN_EXPERT' },
    { link: 'https://launchpad.support.sap.com/#/incident/create', text: 'OPEN_AN_INCIDENT' },
    { link: 'https://answers.sap.com/index.html', text: 'ASK_THE_SAP_COMMUNITY' }
];

/**
 * @returns not solved message element
 */
export default function NotSolvedMessage(): ReactElement {
    return (
        <div id="middle" className="column">
            <h1>{i18next.t('ISSUE_IS_NOT_RESOLVED')}</h1>
            <div id="hr"></div>
            <div className="guided-answer__node__question">
                <p>
                    <strong>{i18next.t('WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED')}</strong>
                </p>
                <p style={{ fontWeight: 400 }}>
                    {i18next.t('THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE')}
                </p>
            </div>
            <div className="guided-answer__node">
                <FocusZone
                    direction={FocusZoneDirection.vertical}
                    className="guided-answer__node"
                    isCircularNavigation={true}>
                    {options.map((btn, i, key) => (
                        <div className="guided-answer__node" key={i}>
                            <a key={i} className="guided-answer__node__edge" href={btn.link} role="button">
                                {i18next.t(btn.text)}
                            </a>
                        </div>
                    ))}
                    <div className="guided-answer__node">
                        <button
                            className="guided-answer__node__edge"
                            onClick={(): void => {
                                actions.goToAllAnswers();
                            }}>
                            {i18next.t('SEARCH_FOR_ANOTHER_GUIDED_ANSWER')}
                        </button>
                    </div>
                </FocusZone>
            </div>
        </div>
    );
}
