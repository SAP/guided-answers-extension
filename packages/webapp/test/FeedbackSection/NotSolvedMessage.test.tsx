import React from 'react';
import { shallow } from 'enzyme';
import { actions } from '../../src/webview/state';
import { initI18n } from '../../src/webview/i18n';
import NotSolvedMessage from '../../src/webview/ui/components/FeedbackSection/NotSolvedMessage/NotSolvedMessage';

jest.mock('../../src/webview/state', () => {
    return {
        actions: {
            goToAllAnswers: jest.fn()
        }
    };
});

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn().mockReturnValue({ guideFeedback: null })
}));

describe('<NotSolvedMessage/>', () => {
    let wrapper: any;
    initI18n();
    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a NotSolvedMessage Component', () => {
        wrapper = shallow(<NotSolvedMessage />);
        const component = wrapper.html();
        expect(component).toMatchInlineSnapshot(
            `"<div id="middle" class="column"><h1>Issue is not resolved</h1><div id="hr"></div><div class="guided-answer__node__question"><p><strong>We are sorry to hear that your issue is not yet resolved.</strong></p><p style="font-weight:400">There are several options for getting further assistance:</p></div><div class="guided-answer__node"><div class="ms-FocusZone css-101 guided-answer__node" data-focuszone-id="FocusZone0"><div class="guided-answer__node"><a class="guided-answer__node__edge" href="https://launchpad.support.sap.com/#/expertchat/create" role="button">Start an Expert Chat</a></div><div class="guided-answer__node"><a class="guided-answer__node__edge" href="https://launchpad.support.sap.com/#/sae" role="button">Schedule an Expert</a></div><div class="guided-answer__node"><a class="guided-answer__node__edge" href="https://launchpad.support.sap.com/#/incident/create" role="button">Open an Incident</a></div><div class="guided-answer__node"><a class="guided-answer__node__edge" href="https://answers.sap.com/index.html" role="button">Ask the SAP Community</a></div><div class="guided-answer__node"><button class="guided-answer__node__edge">Search for another Guided Answer</button></div></div></div></div>"`
        );
    });
});
