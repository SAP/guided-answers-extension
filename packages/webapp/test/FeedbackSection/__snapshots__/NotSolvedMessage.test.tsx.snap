// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotSolvedMessage/> Should render a NotSolvedMessage component 1`] = `
<div>
  <div
    class="column"
    id="middle"
  >
    <h1>
      Issue is not resolved
    </h1>
    <div
      id="hr"
    />
    <div
      class="guided-answer__node__question"
    >
      <p>
        <strong>
          We are sorry to hear that your issue is not yet resolved.
        </strong>
      </p>
      <p
        style="font-weight: 400;"
      >
        There are several options for getting further assistance:
      </p>
    </div>
    <div
      class="guided-answer__node"
    >
      <div
        class="ms-FocusZone css-101 guided-answer__node"
        data-focuszone-id="FocusZone0"
      >
        <div
          class="guided-answer__node"
        >
          <a
            class="guided-answer__node__edge not-solved-link"
            href="https://launchpad.support.sap.com/#/expertchat/create"
            tabindex="0"
          >
            Start an Expert Chat
          </a>
        </div>
        <div
          class="guided-answer__node"
        >
          <a
            class="guided-answer__node__edge not-solved-link"
            href="https://launchpad.support.sap.com/#/sae"
            tabindex="-1"
          >
            Schedule an Expert
          </a>
        </div>
        <div
          class="guided-answer__node"
        >
          <a
            class="guided-answer__node__edge not-solved-link"
            href="https://launchpad.support.sap.com/#/incident/create"
            tabindex="-1"
          >
            Open an Incident
          </a>
        </div>
        <div
          class="guided-answer__node"
        >
          <a
            class="guided-answer__node__edge not-solved-link"
            href="https://answers.sap.com/index.html"
            tabindex="-1"
          >
            Ask the SAP Community
          </a>
        </div>
        <div
          class="guided-answer__node"
        >
          <button
            class="guided-answer__node__edge"
            tabindex="-1"
          >
            Search for another Guided Answer
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
